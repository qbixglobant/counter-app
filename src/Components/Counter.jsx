import { useState } from "react";
import PropTypes from "prop-types";

const Counter = ({ value }) => {
    const h1 = "text-center mt-3 green display-1";
    const h2 = "text-center blue display-2";
    const btn = "btn btn-primary";
    const resetBtn = `${btn} mx-2`;

    const [counter, setCounter] = useState(value);

    const handleIncrement = () => {
        // setCounter( c => c + 10 );
        setCounter(counter + 10);
    };

    const handleReset = () => setCounter(value);

    const handleDecrement = () => {
        (counter > 0) && setCounter(counter - 10);
    };

    return (
        <>
            <h1 className={ h1 }>Counter App</h1>
            <hr />
            <h2 id="counter" className={ h2 }>{ counter }</h2>
            <section className="text-center mt-3">
                <button id="increment" onClick={ handleIncrement } className={ btn }>+</button>
                <button id="reset" onClick={ handleReset }     className={ resetBtn }>reset</button>
                <button id="decrement" onClick={ handleDecrement } className={ btn }>-</button>
            </section>
        </>
    );
};

export default Counter;

Counter.defaultProps = {
    value: 0
};

Counter.propTypes = {
    value: PropTypes.number.isRequired
};