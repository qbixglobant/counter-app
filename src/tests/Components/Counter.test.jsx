import { shallow } from "enzyme";
import Counter from "../../Components/Counter";

describe("Tests on <Counter /> Component", () => {

    let wrapper;

    beforeEach( () => {
        wrapper = shallow(<Counter />);
    });

    test("should match with snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("should match given counter value with component counter value", () => {
        const value = 100;
        const wrapper = shallow(<Counter value={value} />);
        const counterValue = wrapper.find("#counter").text().trim();
        expect(counterValue).toEqual(`${value}`);
    });

    test("should increment counter value by 10", () => {
        // const button = wrapper.find("button").at(0).html();
        wrapper.find("#increment").simulate("click");
        const counterValue = wrapper.find("#counter").text().trim();
        expect(counterValue).toBe( "10" );
    });

    test("should reset to default value", () => {
        const value = 100;
        const wrapper = shallow( <Counter value={ value } /> );
        wrapper.find("#increment").simulate("click");
        let counterValue = wrapper.find("#counter").text().trim();
        expect( counterValue ).toBe( `${value + 10}` );
        wrapper.find("#reset").simulate("click");
        counterValue = wrapper.find("#counter").text().trim();
        expect( counterValue ).toBe( `${value}` );
    });

    test("should decrement counter by 10", () => {
        const value = 50;
        const wrapper = shallow( <Counter value={ value } /> );
        wrapper.find("#decrement").simulate("click");
        const counterValue = wrapper.find("#counter").text().trim();
        expect(counterValue).toBe( "40" );
    });

    test("should not decrement counter if initial value is 0", () => {
        wrapper.find("#decrement").simulate("click");
        const counterValue = wrapper.find("#counter").text().trim();
        expect(counterValue).toBe( "0" );
    });

});
