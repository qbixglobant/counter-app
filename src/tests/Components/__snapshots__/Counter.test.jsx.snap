// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <Counter /> Component should match with snapshot 1`] = `
<Fragment>
  <h1
    className="text-center mt-3 green display-1"
  >
    Counter App
  </h1>
  <hr />
  <h2
    className="text-center blue display-2"
    id="counter"
  >
    0
  </h2>
  <section
    className="text-center mt-3"
  >
    <button
      className="btn btn-primary"
      id="increment"
      onClick={[Function]}
    >
      +
    </button>
    <button
      className="btn btn-primary mx-2"
      id="reset"
      onClick={[Function]}
    >
      reset
    </button>
    <button
      className="btn btn-primary"
      id="decrement"
      onClick={[Function]}
    >
      -
    </button>
  </section>
</Fragment>
`;
